bd-artifact-upload/src/uploader.rs

```diff
@@ -0,0 +1,666 @@
+      .await
+    {
+      log::debug!("failed to write index: {e}");
+    }
+  }
+
+  async fn upload_artifact(
+    data_upload_tx: tokio::sync::mpsc::Sender<DataUpload>,
+    contents: Vec<u8>,
+    name: String,
+  ) -> Result<()> {
+    let path = REPORT_DIRECTORY.join(&name);
+    log::debug!("uploading artifact: {}", path.display());
+
+    // Use exponential backoff to avoid retrying over and over again in case something is going
+    // wrong. We put no overall tiemout as the device might be offline for a long time and we want
+    // to give it whatever time it needs to perform the upload.
+
+
+    let mut retry_backoff = Self::make_retry_policy();
+
+    loop {
+      let upload_uuid = TrackedArtifactUpload::upload_uuid();
+      let (tracked, response) = TrackedArtifactUpload::new(
+        upload_uuid.clone(),
+        UploadArtifactRequest {
+          upload_uuid,
+          type_id: "client_report".to_string(),
+          contents: contents.clone(),
+          artifact_id: name.clone(),
+          ..Default::default()
+        },
+      );
+
+      data_upload_tx
+        .send(DataUpload::ArtifactUpload(tracked))
+        .await
+        .map_err(|_| Error::Shutdown)?;
+
+      if response.await.is_ok_and(|r| r.success) {
+        log::debug!("upload of artifact: {name} succeeded");
+        break;
+      }
+
+      let delay = retry_backoff.next_backoff().unwrap();
+
+      log::debug!("upload of artifact: {name} failed, retrying in {delay:?}");
+      tokio::time::sleep(delay).await;
+    }
+
+    Ok(())
+  }
+
+  async fn perform_intent_negotiation(
+    data_upload_tx: tokio::sync::mpsc::Sender<DataUpload>,
+    id: String,
+  ) -> Result<IntentDecision> {
+    let mut retry_backoff = Self::make_retry_policy();
+
+    loop {
+      let upload_uuid = TrackedArtifactIntent::upload_uuid();
+      let (tracked, response) = TrackedArtifactIntent::new(
+        upload_uuid.clone(),
+        UploadArtifactIntentRequest {
+          type_id: "client_report".to_string(),
+          artifact_id: id.to_string(),
+          intent_uuid: upload_uuid.to_string(),
+          // TODO(snowp): Figure out how to send relevant metadata about the artifact here.
+          metadata: vec![],
+          ..Default::default()
+        },
+      );
+
+      data_upload_tx
+        .send(DataUpload::ArtifactUploadIntent(tracked))
+        .await
+        .map_err(|_| Error::Shutdown)?;
+
+      if let Ok(response) = response.await {
+        break Ok(response.decision);
+      }
+
+      let delay = retry_backoff.next_backoff().unwrap();
+      log::debug!("intent negotiation for artifact: {id} failed, retrying in {delay:?}");
+      tokio::time::sleep(delay).await;
+    }
+  }
+
+  fn make_retry_policy() -> ExponentialBackoff<SystemClock> {
+    // TODO(snowp): Make more of these parameters runtime configurable.
+
+    // This default sets an initial timeout of 500s with a max timeout of 1 min (jittered).
```

Comment:
```suggestion
    // This default sets an initial timeout of 500ms with a max timeout of 1 min (jittered).
```

I assume? For this one can we just share whatever backoff policy we are using in API? I think I'm using much longer settings there but I can't remember. We can move that backoff policy to client-common?

Reply:
The API one seems to use 30s/30min but has their own runtime flags that drive them - I wonder if we should consolidate the runtime flags and just have one configuration for general backoff or if it makes sense to keep them separate?

Reply:
My vote would be to just consolidate.